### ISSUE TITLE
/sdcard/Android/data inaccessible on Android 11

### ISSUE THREAD

**Issue by calisro** [pacien/tincapp#103]
The documentation states that the data directory will be accessible on Android 11 via the standard file manager or adb however on my Pixel 4a 5g it is not.

bramble:/ $ cd /sdcard/Android/data/*tin*
/system/bin/sh: cd: /sdcard/Android/data/*tin*: No such file or directory
2|bramble:/ $ cd /sdcard/Android/data/                                                                                                           
bramble:/sdcard/Android/data $ ls
ls: .: Permission denied
1|bramble:/sdcard/Android/data $ ls- ^C
130|bramble:/sdcard/Android/data $ ls -al
ls: .: Permission denied

Without being able to access that directory, there is no way to customize tinc.  Can we have a way to configure the path to these or some other way of importing the configs?  Thank-you!

**Comment by pacien** [comment]
The parent directory isn't always listable.
You can perhaps try typing out the full path with `org.pacien.tincapp` at the end.

**Comment by calisro** [comment]
127|bramble:/sdcard/Android/data $ ls org.pacien.tincapp                                                                                         
ls: org.pacien.tincapp: Permission denied
1|bramble:/sdcard/Android/data $ ls org.pacien.tincapp/files/                                                                                    
ls: org.pacien.tincapp/files/: Permission denied
1|bramble:/sdcard/Android/data $ ls /storage/emulated/0/Android/data/org.pacien.tincapp/files/
ls: /storage/emulated/0/Android/data/org.pacien.tincapp/files/: Permission denied
1|bramble:/sdcard/Android/data $ ls /storage/emulated/0/Android/data/org.pacien.tincapp/                                                         
ls: /storage/emulated/0/Android/data/org.pacien.tincapp/: Permission denied
1|bramble:/sdcard/Android/data $ ls /sdcard/Android/data/org.pacien.tincapp/                                                                     
ls: /sdcard/Android/data/org.pacien.tincapp/: Permission denied
1|bramble:/sdcard/Android/data $

**Comment by pacien** [comment]
Could you also try with `/storage/sdcard0/Android/data/org.pacien.tincapp`? Same with `sdcard1` instead of `sdcard0`?

`#tincapp:pacien.net` on Matrix or `#tincapp` on `irc.freenode.net` might be a better place to sort this out.

**Comment by calisro** [comment]
adb shell
bramble:/ $ cd /storage/sdcard0/Android/data/org.pacien.tincapp
/system/bin/sh: cd: /storage/sdcard0/Android/data/org.pacien.tincapp: No such file or directory
d /storage/sdcard1/Android/data/org.pacien.tincapp                            <
/system/bin/sh: cd: /storage/sdcard1/Android/data/org.pacien.tincapp: No such file or directory
2|bramble:/ $                                                                                                                                         

Those don't exist.  This isn't a location issue.  Android 11 specifically denies access to the sdcard data directories except for media.  Without changes to allow us to use a different location, there will soon be no way to use this app as a person running stock android won't be able to edit configs.

https://developer.android.com/about/versions/11/privacy/storage

**Comment by pacien** [comment]
I had previously observed that this directory remained accessible through ADB on some devices and on the emulator despite them running Android 11. Since this doesn't seem to be the case everywhere, it indeed becomes necessary to find an alternative way to make the configuration directory accessible.

Here are a few ideas that have been suggested by users so far:
* Add a file explorer and file editor within the application. This is quite heavy and complex to implement and I couldn't find any drop-in library that could be used to make that easy.
* Expose the tinc control socket to the network and use tinc commands to alter the configuration. This requires starting the tinc daemon before having the network interface configured, changing a lot in the application's logic. This is probably a dead end since the network interface configuration file (network.conf) also wouldn't be editable, since it's not known by tinc itself.
* Allow exporting and importing the whole configuration directory as some tar file written and read to an accessible location like the Downloads folder. This is feasible but highly inconvenient to simply edit the files and error prone if importing some archive overwrites or deletes important keys.
* Expose the configuration directory through FTP or SFTP (without full SSH). This would allow users to browse the configuration directory and edit files from both a remote machine or through a local FTP or SFTP client on Android, many of which even include a text editor. Multiple libraries seem to offer pure Java implementations of FTP and SFTP servers, making this not too hard to implement.

Other ideas are welcome. For now, I'm leaning towards the last solution above.

**Comment by pacien** [comment]
I've experimented with alternative configuration paths, with no success. I ended up implementing the FTP-based solution to make the configuration directory accessible. (I couldn't get an embedded SSH/SFTP server to work without breaking compatibility with older Android versions, so I went for FTP instead).

The FTP server can be activated and deactivated on the configuration page. The connection endpoint and randomly generated credentials are displayed there afterwards. It is then possible to access the tinc network configuration directory in `files/networks` through any FTP client such as FileZilla on desktop or Ghost Commander on Android.

[Here][apk]'s the preview (debug) APK. Because it is signed with a debug key, you'll need to uninstall any production version that you already have on your device, after making a backup of your current configuration. After testing, you'll also need to manually uninstall this debug version to be able to receive future updates through Google Play and F-Droid.

[apk]: https://files.pacien.net/tmp/tincapp.0.31-android11ftpconfigaccess.debug.apk

@calisro Could you please test this?

If everything works as expected, I'll make a new release and update the documentation accordingly.

**Comment by calisro** [comment]
@pacien  I should be able to test this out soon.  Appreciate all the efforts.

**Comment by calisro** [comment]
Some feedback.  Can you **please** make the password/port at a minimum configurable?  Any tool I use, i'd want to be able to store the connection so I can quickly edit these files without retyping every time a new password.  The password seemed to change on me a few times.  

As far as usability, it works as expected for typical use. :)   I did run into some problems with using a multi-threaded tool against it as seen below.

```
rclone copy files/ tinc4a5g:files/networks/ -vv
2020/12/08 15:38:40 DEBUG : rclone: Version "v1.54.0-beta.4893.ceeac84cf" starting with parameters ["rclone" "copy" "files/" "tinc4a5g:files/networks/" "-vv"]
2020/12/08 15:38:40 DEBUG : Creating backend with remote "files/"
2020/12/08 15:38:40 DEBUG : Using config file from "/home/xxx/.rclone.conf"
2020/12/08 15:38:40 DEBUG : fs cache: renaming cache item "files/" to be canonical "/home/xxx/tincstuff/files"
2020/12/08 15:38:40 DEBUG : Creating backend with remote "tinc4a5g:files/networks/"
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Connecting to FTP server
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Connecting to FTP server
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Waiting for checks to finish
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Connecting to FTP server
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Waiting for transfers to finish
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Connecting to FTP server
2020/12/08 15:38:40 ERROR : ftp://10.100.1.129:65521/files/networks: Error while Logging in into 10.100.1.129:65521: EOF
**2020/12/08 15:38:40 DEBUG : hometun/network.conf: Received error: Put mkParentDir failed: mkdir "files/networks/hometun" failed: getInfo: ftpConnection Login: EOF - low level retry 1/10**
2020/12/08 15:38:40 DEBUG : ftp://10.100.1.129:65521/files/networks: Connecting to FTP server
2020/12/08 15:38:41 INFO  : hometun/ed25519_key.priv: Copied (new)
2020/12/08 15:38:41 INFO  : hometun/network.conf1: Copied (new)
2020/12/08 15:38:41 INFO  : hometun/network.conf: Copied (new)
2020/12/08 15:38:41 INFO  : hometun/tinc.conf: Copied (new)
2020/12/08 15:38:41 INFO  : hometun/hosts/home: Copied (new)
2020/12/08 15:38:41 INFO  : hometun/hosts/p31: Copied (new)
2020/12/08 15:38:41 ERROR : hometun/hosts/p3: Failed to copy: update getinfo: short response: 226 Closing data connection.
```

I am getting intermittent errors on the ftp server.  It eventually crashed.  I've attached the generated log with traces to help debug.  

I'm using rclone to transfer and the crash/errors could very well be the multi-treaded nature of rclone (default is 4 threads).  But thought it might be good from a testing perspective.

[log.txt](https://github.com/pacien/tincapp/files/5661865/log.txt)

Also note that if I run rclone single threaded I didn't get any errors but there may be a race condition or something in the ftp implementation. If i use 1 thread I didn't get errors but as soon as I turn it to more than 1, it starts erroring intermittantly.  Not sure if this is worth looking into.

**Comment by pacien** [comment]
> Can you please make the password/port at a minimum configurable?
> […]
> The password seemed to change on me a few times.

The password is now stable and only randomly initialised the first time.
The credentials and ports can also be modified in a preference file.

>  I did run into some problems with using a multi-threaded tool against it as seen below.

This should be solved as well.
I've simply disabled multi-threading in the FTP library, which was overkill for the usage in that app.

**Comment by calisro** [comment]
> The password is now stable and only randomly initialised the first time.
> The credentials and ports can also be modified in a preference file.

thats great.

and yes agree with disabling the multi-threading.  

Is there another apk you want me to test?  or you think its fine.

**Comment by pacien** [comment]
I've already tested the fix with rclone as you have described, so it should be fine :+1: 
The new release is already published as `v0.32` and should be soon deployed through Google Play and F-Droid.
You'll need to make a backup of your current configuration and uninstall the debug APK to get the update.

**Comment by platise** [comment]
I am struggling to edit files on Android 13:

- adb: can read but no write permissions, cannot go to adb root unless I root the phone
- usb storage mode, can only read
- ssh server: same, ready only

Could you return the ftp feature that was removed not that long ago:

## v0.38 -- 2023-07-30
- make configuration files and logs accessible in the user-accessible storage
  (in USB storage mode). The embedded FTP server has been removed
